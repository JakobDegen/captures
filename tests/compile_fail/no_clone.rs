use captures::*;

// Check the basic error message on a missing `Clone` impl
fn no_generic() {
    struct S;
    let s = S;

    let f = capture! (
        clone s,
        move || {
            let l = s;
            l
        }
    );
    f();
}

// Check the error message when the type is generic
fn no_derive() {
    struct S<T>(T);
    let s = S(0);

    let f = capture! (
        clone s,
        move || {
            let l = s;
            l
        }
    );
    f();
}

// Check the error message when only the generic parameter is missing the impl
fn no_impl() {
    struct NonClone;

    #[derive(Clone)]
    struct S<T>(T);
    let s = S(NonClone);

    let f = capture! (
        clone s,
        move || {
            let l = s;
            l
        }
    );
    f();
}

fn main() {
    no_generic();
    no_derive();
    no_impl();
}
