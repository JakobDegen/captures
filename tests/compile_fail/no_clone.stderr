error[E0277]: the trait bound `no_generic::S: Clone` is not satisfied
   --> tests/compile_fail/no_clone.rs:9:9
    |
9   |         clone s,
    |         ^^^^^^^ the trait `Clone` is not implemented for `no_generic::S`
    |
note: required by `clone`
   --> $RUST/core/src/clone.rs
    |
    |     fn clone(&self) -> Self;
    |     ^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `no_derive::S<{integer}>: Clone` is not satisfied
   --> tests/compile_fail/no_clone.rs:24:9
    |
24  |         clone s,
    |         ^^^^^^^ the trait `Clone` is not implemented for `no_derive::S<{integer}>`
    |
note: required by `clone`
   --> $RUST/core/src/clone.rs
    |
    |     fn clone(&self) -> Self;
    |     ^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `NonClone: Clone` is not satisfied
   --> tests/compile_fail/no_clone.rs:42:9
    |
42  |         clone s,
    |         ^^^^^^^
    |         |
    |         expected an implementor of trait `Clone`
    |         help: consider borrowing here: `&clone s`
    |
note: required because of the requirements on the impl of `Clone` for `no_impl::S<NonClone>`
   --> tests/compile_fail/no_clone.rs:37:14
    |
37  |     #[derive(Clone)]
    |              ^^^^^
note: required by `clone`
   --> $RUST/core/src/clone.rs
    |
    |     fn clone(&self) -> Self;
    |     ^^^^^^^^^^^^^^^^^^^^^^^^
    = note: this error originates in the derive macro `Clone` (in Nightly builds, run with -Z macro-backtrace for more info)
